Leehandsub

leehandsub opened the pull request:
@rlarlvy153 잘 부탁드립니다!

leehandsub committed:
docs : README 작성

- 기능 목록
- 추가된 요구 사항 정리

leehandsub committed:
feat : 이름 설정 구현

leehandsub committed:
feat : 횟수 설정 구현

leehandsub committed:
feat : 횟수와 이름을 같이 관리하기 위한 data class 추가

leehandsub committed:
feat : 각각에 대한 전진

- string 값을 racingcar 클래스로 변환
- racingcar 정보의 거리를 랜덤으로 plus

leehandsub committed:
feat : 횟수만큼 반복하게 구현

- for문을 사용해서 gamecount만큼 반복

leehandsub committed:
feat : 전진을 4이상할 경우 전진

- 하드 코딩된 부분 수정
- 조건에 따른 거리 추가

leehandsub committed:
feat : 게임 진행 중 대한 출력

- outputview 구현
- racingCar 안의 name, distance로 판단해서 출력

leehandsub committed:
feat : 우승자에 대한 결과 계산

- 여러 명일수도 있기 때문에 list로 return

leehandsub committed:
feat : 우승자에 대한 결과 출력

- jointostring으로 마지막 추가하지 않기

leehandsub committed:
feat : 이름은 5자 이하로 구현

leehandsub committed:
feat : 주어진 횟수가 숫자가 아닐 경우

leehandsub committed:
fix : indent 3이 넘지 않도록 수정

- for문안에 잇는 if문을 함수로 빼서 구현

rlarlvy153 commented on docs/README.md:
```diff
@@ -0,0 +1,38 @@
+5. 추가사항
+    - [x]  커밋 컨벤션을 지켰는가?
+    - [x]  indent(인덴트, 들여쓰기) depth를 3이 넘지 않도록 구현
+    - [ ]  기능 목록이 정상 동작함을 테스트 코드로 확인
```
이건 왜 덜했니?

leehandsub replied:
test 코드 어떻게 해야할지 모르겟어서 하하...

rlarlvy153 commented on src/main/kotlin/racingcar/controller/RacingCarController.kt:
```diff
@@ -0,0 +1,78 @@
+
+    fun playGame() {
+        inputView.printInputCarName()
+        val userList = Console.readLine().split(NAME_DELIMITERS)
```
인풋뷰에서 해야될거같다

rlarlvy153 replied:
밑에것들도 인풋뷰에서 해야될것들이 있네

rlarlvy153 commented on src/main/kotlin/racingcar/controller/RacingCarController.kt:
```diff
@@ -0,0 +1,78 @@
+        val gameCount = Console.readLine()
+        checkNumber(gameCount)
+
+        racingCarList = gameContinue(racingCarList, gameCount.toInt())
```
toInt 이런거 인풋뷰에서 다 검증해서 올리기

rlarlvy153 commented on src/main/kotlin/racingcar/controller/RacingCarController.kt:
```diff
@@ -0,0 +1,78 @@
+        outputView.printResult()
+
+        for (i in 0 until gameCount) {
+            racingCarList.map {
```
걍 돌기만 하는데 forEach가 나아보임
map 리턴값을 안쓰

rlarlvy153 commented on src/main/kotlin/racingcar/controller/RacingCarController.kt:
```diff
@@ -0,0 +1,78 @@
+    private fun gameContinue(racingCarList: List<RacingCar>, gameCount: Int): List<RacingCar> {
+        outputView.printResult()
+
+        for (i in 0 until gameCount) {
```
i안쓰는데, repeat(gameCount) { } 사용해보삼

rlarlvy153 commented on src/main/kotlin/racingcar/controller/RacingCarController.kt:
```diff
@@ -0,0 +1,78 @@
+            outputView.printEnter()
+        }
+
+        return racingCarList
```
얘는 인풋이랑 리턴이랑 차이가 없는데 왜있음?? 레이싱카리스트는 함수 중간에 변경되지 않는데

leehandsub replied:
그 racingcar 각각의 값을  addRacingCarDistance를 통해서 변화 시켜요

rlarlvy153 replied:
그럼 map이 아니어야될거같은데 map은 형변환이잖아

leehandsub replied:
네넹 그래서 forEach로 바꿨습니다!

rlarlvy153 commented on src/main/kotlin/racingcar/controller/RacingCarController.kt:
```diff
@@ -0,0 +1,78 @@
+        outputView.printMaxDistanceRacingCarName(maxDistanceRacingCarList)
+    }
+
+    private fun addRacingCarDistance(racingCar: RacingCar): RacingCar {
```
함수내에서 distnace만 쓰는거같은데 레이싱카를 받아야됨?

leehandsub replied:
그 addRacingCarDistance에서 racingCar를 출력해서 그런데 addRacingCarDistance에서 출력하게 하면 안되는거겟죠?

rlarlvy153 replied:
add어쩌구 함수인데 레이싱카를 리턴함
네이밍과 리턴이 안맞음

rlarlvy153 commented on src/main/kotlin/racingcar/controller/RacingCarController.kt:
```diff
@@ -0,0 +1,78 @@
+        return racingCar
+    }
+
+    private fun findMaxDistanceRacingCar(racingCarList: List<RacingCar>): List<String> {
```
네이밍에 자료형을 포함하지 않기
racingCarList -> rackingCars
userList -> users 
그리고 일급 컬렉션이란걸 찾아보기

leehandsub replied:
일급 컬렉션 요부분은 조금 더 공부를 해보겠습니다. 감은 조금 오는데 어떻게 적용해야할지를 모르겠네요

rlarlvy153 commented on src/main/kotlin/racingcar/controller/RacingCarController.kt:
```diff
@@ -0,0 +1,78 @@
+            .map { it.racingCarName }
+    }
+
+    private fun checkSize(userList: List<String>) {
```
이름너무 추상적임
유저 개수 체크 같은.. 함수 내용 구체적으로 알 수 있게 네이밍

rlarlvy153 commented on src/main/kotlin/racingcar/controller/RacingCarController.kt:
```diff
@@ -0,0 +1,78 @@
+    }
+
+    companion object {
+        const val RANDOM_START_NUM = 0
```
여기서만 쓰는거같은데 private 해도될듯

rlarlvy153 commented on src/main/kotlin/racingcar/model/RacingCar.kt:
```diff
@@ -0,0 +1,3 @@
+package racingcar.model
+
+data class RacingCar(val racingCarName: String, var distance: Int)
```
var는 노출하지말고 함수로 다룰 수 있는지 확인

leehandsub replied:
어떤 느낌으로 해야할지를 잘모르겟네요 해본 방법 
```
data class RacingCar(val racingCarName: String, val distance: Int){
    fun distanceIncrease(randomNum : Int) = RacingCar(racingCarName,distance+randomNum )
}
```
```
private fun addRacingCarDistance(racingCar: RacingCar): RacingCar {
      ...
            racingCar.copy(racingCarName = racingCar.racingCarName,racingCar.distance+randomNum)
      ...
    }
```
첫번째,두번째 방법 둘다 racingCar는 불변해서 결국 객체가  새로 바뀔수가 없기 때문에 값이 안바뀌는 어떻게 해야할지 감이 안잡히네요

rlarlvy153 replied:
private var 하고
fun increase 어쩌구 하란 뜻이엇는디

rlarlvy153 commented on src/main/kotlin/racingcar/view/InputView.kt:
```diff
@@ -0,0 +1,11 @@
+    fun printInputGameCount() = println(INPUT_GAME_COUNT_MESSAGE)
+
+    companion object {
+        const val INPUT_CAR_NAME_MESSAGE = "경주할 자동차 이름을 입력하세요.(이름은 쉼표(,) 기준으로 구분)"
```
얘도 private

rlarlvy153 commented on src/main/kotlin/racingcar/view/OutputView.kt:
```diff
@@ -0,0 +1,30 @@
+    private fun printDistance(distance: Int): String {
+        var temp = ""
+        for (i in 0 until distance) {
+            temp += DISTANCE_EXPRESS
```
StringBuilder, StringBuffer 차이 알아보고, 적절한거 사용하기
스트링 직접붙이는거 피하기

rlarlvy153 commented on src/main/kotlin/racingcar/view/OutputView.kt:
```diff
@@ -0,0 +1,30 @@
+
+    private fun printDistance(distance: Int): String {
+        var temp = ""
+        for (i in 0 until distance) {
```
안에서 repeat() 쓰기

rlarlvy153 commented on src/main/kotlin/racingcar/view/OutputView.kt:
```diff
@@ -0,0 +1,30 @@
+    }
+
+    fun printMaxDistanceRacingCarName(maxDistanceRacingCarList: List<String>) =
+        println("$FINAL_WINNER_MESSAGE${maxDistanceRacingCarList.joinToString(", ")}")
```
밑에 콤마 상수로 선언해놓고 여기에 하드코딩해놨네

leehandsub committed:
fix : inputview에서 해야할 일이 controller에서 구현 수정

- input에 대한 검증과, 데이터 가공을 inputview에서 구현

leehandsub committed:
fix : 올바르지 않는 내부 함수 수정

- map을 쓰지만 안에 내용을 쓰지 않기 때문에 forEach로 변경

leehandsub committed:
fix : i를 쓰지 않끼 때문에 kotlin 내부 함수 사용

- repeart으로 대체

leehandsub committed:
fix : 변수명에 list넣지 않기

- list -> s로 변경

leehandsub committed:
fix : 애매한 네이밍 수정

- 그냥 사이즈말고 레이싱카 이름 사이즈 크기 확인으로 변경

leehandsub committed:
fix : private일거 같은 변수 변경

- 추후 다른데서 쓸 수 있는 것은 변경 안함

leehandsub committed:
fix : stringbulder를 이용해서 문자열 구현

- 그냥 += 할경우 그전께 그냥 남아만있고 주소가 없어져 쓸데 없는 값이 있는 것을 알게 됨

leehandsub commented:
@rlarlvy153 조금 고치긴 했는데 질문사항 같은거는 일단 comment로 올렸습니다.

rlarlvy153 commented on src/main/kotlin/racingcar/controller/RacingCarController.kt:
```diff
@@ -0,0 +1,57 @@
+
+        repeat(gameCount) {
+            rackingCars.forEach {
+                addRacingCarDistance(it)
```
이거 it 겹친다고 워닝뜨지싶은데 네이밍 주삼

rlarlvy153 commented on src/main/kotlin/racingcar/controller/RacingCarController.kt:
```diff
@@ -0,0 +1,57 @@
+            outputView.printEnter()
+        }
+
+        return rackingCars
```
함수이름이 게임 컨티뉴인데 카를 리턴한다 의미가 안맞는듯
게다가 오타있음

rlarlvy153 commented on src/main/kotlin/racingcar/controller/RacingCarController.kt:
```diff
@@ -0,0 +1,57 @@
+        return rackingCars
+    }
+
+    private fun gameResult(racingCarList: List<RacingCar>) {
```
게임 리절트는 뭔가 리절트를 리턴해야될거같은 네이밍인데 암것도 리턴안하고 걍 종료출력만하는걸로 보이는데, 그럼 gameEnd 라거나.. 뭐 gameOver나 바꿔야될듯

rlarlvy153 commented on src/main/kotlin/racingcar/controller/RacingCarController.kt:
```diff
@@ -0,0 +1,57 @@
+    private fun addRacingCarDistance(racingCar: RacingCar): RacingCar {
+        val randomNum = Randoms.pickNumberInRange(RANDOM_START_NUM, RANDOM_FINISH_NUM)
+        if (randomNum >= MIN_DISTANCE_NUM) {
+            racingCar.distance += randomNum
```
레인지만 카 객체에 전달하고
distance는 private로 두고 카 내에서 다뤄야할걸로 보임
자동차 거리가 증가하는건 컨트롤러가 할일이 아니고 차 각각의 파라미터니까 차 내에서 처리

rlarlvy153 commented on src/main/kotlin/racingcar/controller/RacingCarController.kt:
```diff
@@ -0,0 +1,57 @@
+    }
+
+    private fun findMaxDistanceRacingCar(racingCarList: List<RacingCar>): List<String> {
+        return racingCarList.filter { racingCar -> racingCar.distance == racingCarList.maxOf { it.distance } }
```
class RacingCars  {
 private val racingCars = List<Car>
 
fun getMaxDistanceCar() {
return racingCarList.filter { racingCar -> racingCar.distance == racingCarList.maxOf { it.distance } }
}
}
이런식으로 하는게 일급컬렉션임
컨트롤러가 지금은 레이싱카의 구현체를 너무 잘 알고있음 안으로 숨기기

leehandsub replied:
일단 getMaxDistanceCar에서만 쓰긴했는데 일급 컬렉션 쓰면 모든 List<RacingCar>를 RacingCars로 바꿔야하나요?

rlarlvy153 commented on src/main/kotlin/racingcar/view/InputView.kt:
```diff
@@ -0,0 +1,38 @@
+import racingcar.model.RacingCar
+
+class InputView {
+    fun printInputCarName(): List<RacingCar> {
```
인풋뷰가 레이싱카라는 비즈니스로직의 객체를 알고있는데
몰라야될거같음
UI만 딱 받아서 컨트롤러에 넘기고, 컨트롤러는 비즈니스로직을 알아도 되기 떄문에 거기서 RacingCar 생성

rlarlvy153 replied:
얘도 프린트 함수인데 인트리턴함

함수가 여러가지역할을 하면안됨 1줄만 들어가더라도 무조건 한가지 역할만

rlarlvy153 commented on src/main/kotlin/racingcar/view/InputView.kt:
```diff
@@ -0,0 +1,38 @@
+        return users.map { RacingCar(it, INIT_DISTANCE_NUM) }
+    }
+
+    fun printInputGameCount(): Int {
```
프린트인데 인트 리턴함...

rlarlvy153 commented on src/main/kotlin/racingcar/view/InputView.kt:
```diff
@@ -0,0 +1,38 @@
+    }
+
+    private fun checkNumber(input: String) {
+        require(input.toIntOrNull() != null)
```
requireNotNull 알아보기

rlarlvy153 replied:
여기서 toIntOrNull 하면 이거 밖에서 toInt 또하는데 비효율적임
한번만 할 수 있게 구현

rlarlvy153 commented on src/main/kotlin/racingcar/view/InputView.kt:
```diff
@@ -0,0 +1,38 @@
+        private const val NAME_DELIMITERS = ","
+        const val INIT_DISTANCE_NUM = 0
+        private const val INPUT_CAR_NAME_MESSAGE = "경주할 자동차 이름을 입력하세요.(이름은 쉼표(,) 기준으로 구분)"
+        private const val INPUT_GAME_COUNT_MESSAGE = "시도할 횟수는 몇 회인가요?"
```
private이랑 non private이랑 섞여있음

눈으로 보기에 구분될 수 있게 몰아주든지, 역할별로 구분해서 엔터쳐주든지

leehandsub committed:
fix : controller에서 많은 역할을 하는것을 수정

- controller에서 레이싱카에 대한 일을 하는 거 같아 이것을 racingcar class하도록 수정
- var을 public하게 쓰는것보다 private으로 쓰는게 좋음

leehandsub committed:
fix : 함수이름과 리턴이 안 맞아 수정

- continue인데 리턴하는 것을 get으로 변경
- result지만 리턴을 안해서 end로 변경

leehandsub committed:
fix : inputview에서 비지니스 로직 처리하는 부분 삭제

- 비지니스 로직 구현체는 controller에 보냄
- 오타 수정

leehandsub committed:
fix : toInt를 두번하는 로직 정리

- 함수이름과 맞게 구현
- 최대한 기능을 한개만 하게 구현
- toInt를 두번안하고 숫자인지 체크할때 toInt바꿔서 바로 return하게 구현

leehandsub committed:
fix : 일급컬렉션 사용해서 racingcars에 대한 것을 컨트롤러에서 숨기기

- racingcars를 객체화

leehandsub commented:
@rlarlvy153 리뷰 최대한 반영했습니다!

rlarlvy153 commented on src/main/kotlin/racingcar/controller/RacingCarController.kt:
```diff
@@ -0,0 +1,56 @@
+        gameEnd(racingCars)
+    }
+
+    private fun convertUsersToRacingCars(users: List<String>): List<RacingCar> =
```
RacingCars가 있는데 여긴 왜 List<RacingCar>를 사용하너

rlarlvy153 commented on src/main/kotlin/racingcar/controller/RacingCarController.kt:
```diff
@@ -0,0 +1,56 @@
+        users.map { RacingCar(it, INIT_DISTANCE_NUM) }
+
+
+    private fun getGameResultRacingCars(racingCars: List<RacingCar>, gameCount: Int): List<RacingCar> {
```
여기도 List쓰지말기

rlarlvy153 commented on src/main/kotlin/racingcar/controller/RacingCarController.kt:
```diff
@@ -0,0 +1,56 @@
+    }
+
+    private fun findMaxDistanceRacingCar(racingCarList: List<RacingCar>): List<String> {
+        return RacingCars(racingCarList).getMaxDistanceCar().map { it.racingCarName }
```
개수만 넘기고 안에서생성
